src/clients/SpokePoolClient.ts

```diff
@@ -239,6 +245,63 @@ export class SpokePoolClient {
     return `${event.depositId}-${event.originChainId}`;
   }
 
+  // TODO: Should unit test this binary search algorithm.
+  // Look for the block number of the event that emitted the deposit with the target deposit ID. We know that
+  // `numberOfDeposits` is strictly increasing for any SpokePool, so we can use a binary search to find the blockTag
+  // where `numberOfDeposits == targetDepositId`.
+  async binarySearchForBlockContainingDepositId(targetDepositId: number): Promise<number> {
+    let high = this.firstBlockToSearch;
+    let low = this.spokePoolDeploymentBlock;
+    do {
+      const mid = Math.floor((high + low) / 2);
+      const searchedDepositId = await this.spokePool.numberOfDeposits({ blockTag: mid });
+      if (targetDepositId > searchedDepositId) low = mid + 1;
+      else if (targetDepositId < searchedDepositId) high = mid - 1;
+      else return mid;
+    } while (low <= high);
+  }
+
+  // Load a deposit for a fill if the fill's deposit ID is less than this client's earliest searched
+  // deposit ID. This can be used by the Dataworker to determine whether to give a relayer a refund for a fill
+  // of a deposit older than its fixed lookback.
+  async queryHistoricalDepositForFill(fill: Fill): Promise<DepositWithBlock | undefined> {
+    if (fill.originChainId !== this.chainId) throw new Error("fill.originChainId !== this.chainid");
+    if (fill.depositId >= this.earliestDepositId) return this.getDepositForFill(fill);
+
+    // Binary search between spoke pool deployment block and earliest block searched to find the block range containing
+    // the deposited event, where numberOfDeposits incremented from depositId-1 to depositId.
+    const [blockBeforeDeposit, blockAfterDeposit] = await Promise.all([
+      this.binarySearchForBlockContainingDepositId(fill.depositId - 1),
+      this.binarySearchForBlockContainingDepositId(fill.depositId + 1),
+    ]);
+
+    const query = await paginatedEventQuery(this.spokePool, this._queryableEventNames().FundsDeposited, {
+      fromBlock: blockBeforeDeposit,
+      toBlock: blockAfterDeposit,
+      maxBlockLookBack: 0,
+    });
+    const event: FundsDepositedEvent = (query as FundsDepositedEvent[]).find(
+      (deposit) => deposit.args.depositId === fill.depositId
+    );
```

Comment:
If you use a topic filter, the node should do the filtering for us.

```suggestion
    const [event] = await paginatedEventQuery(this.spokePool, this.spokePool.filters.FundsDeposited(null, null, null, null, fill.depositId), {
      fromBlock: blockBeforeDeposit,
      toBlock: blockAfterDeposit,
      maxBlockLookBack: 0,
    });
```

Reply:
Done